### PR TITLE
Remove deprecations in CRUDController

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

Remove deprecations found in CRUDController

I am targeting this branch, because this is not BC.

## Changelog

```markdown
### Changed
- Bump exporter to 2.1
### Removed
- `$id` parameter in actions of `CRUDController`
- `sonata.admin.exporter` service
```
I was wondering if it's better to leave the action with empty parameters or add `Request $request`, I'd add the request.

### CODE REVIEW

**Comment by greg0ire** [comment]
> I was wondering if it's better to leave the action with empty parameters or add Request $request, I'd add the request.

I would too.

**Comment by franmomu** [comment]
> > I was wondering if it's better to leave the action with empty parameters or add Request $request, I'd add the request.
> 
> I would too.

Let me change it

File: src/Controller/CRUDController.php [hunk 1]
```diff
@@ -655,13 +622,9 @@ public function showAction($deprecatedId = null) // NEXT_MAJOR: Remove the unuse
...
 
-        // NEXT_MAJOR: replace deprecation with exception
         if (!\is_array($fields->getElements()) || 0 === $fields->count()) {
-            @trigger_error(
-                'Calling this method without implementing "configureShowFields"'
-                .' is not supported since sonata-project/admin-bundle 3.40.0'
-                .' and will no longer be possible in 4.0',
-                E_USER_DEPRECATED
+            throw new \RuntimeException(
+                'No fields defined. Did you forget to implement the "configureShowFields" method?'
```
Comment (phansys) [inline src/Controller/CRUDController.php]: I don't know if this is completely right, but I didn't found this method declared in other interface:
```suggestion
                'No fields defined. Did you forget to implement `AdminExtensionInterface::configureShowFields()`?'
```

File: src/Controller/CRUDController.php [hunk 2]
```diff
@@ -804,11 +751,10 @@ public function historyViewRevisionAction($id = null, $revision = null) // NEXT_
      *
      * @return Response
      */
-    public function historyCompareRevisionsAction($id = null, $base_revision = null, $compare_revision = null) // NEXT_MAJOR: Remove the unused $id parameter
+    public function historyCompareRevisionsAction(Request $request, $base_revision = null, $compare_revision = null)
```
Comment (phansys) [inline src/Controller/CRUDController.php]: Could we leverage these changes to use camel case in argument names? Is required to keep them with snake case?

File: src/Controller/CRUDController.php [hunk 3]
```diff
@@ -1491,23 +1405,17 @@ private function checkParentChildAssociation(Request $request, $object): void
...
 
         if ($parentAdmin->getObject($parentId) !== $propertyAccessor->getValue($object, $propertyPath)) {
-            // NEXT_MAJOR: make this exception
-            @trigger_error(
-                "Accessing a child that isn't connected to a given parent is"
-                ." deprecated since sonata-project/admin-bundle 3.34 and won't be allowed in 4.0.",
-                E_USER_DEPRECATED
-            );
+            throw new \RuntimeException(sprintf(
+                'There is no association between %s and %s',
```
Comment (phansys) [inline src/Controller/CRUDController.php]:
```suggestion
                'There is no association between "%s" and "%s"',
```

File: tests/Controller/CRUDControllerTest.php [hunk 4]
```diff
@@ -2476,19 +2451,8 @@ public function testCreateActionAjaxErrorWithoutAcceptApplicationJson(): void
             ->method('createView')
             ->willReturn($formView);
 
-        $this->expectTranslate('flash_create_error', [
-            '%name%' => '',
-        ], 'SonataAdminBundle');
-
-        $this->assertInstanceOf(Response::class, $response = $this->controller->createAction());
-        $this->assertSame($this->admin, $this->parameters['admin']);
```
Comment (phansys) [inline tests/Controller/CRUDControllerTest.php]: Why these expectations were removed?

File: src/Controller/CRUDController.php (hunk 2 above)
Comment (franmomu) [reply src/Controller/CRUDController.php]: I'll change them

File: tests/Controller/CRUDControllerTest.php (hunk 4 above)
Comment (franmomu) [reply tests/Controller/CRUDControllerTest.php]: Becase [`handleXmlHttpRequestErrorResponse` used to return `null` when the `Accept` parameter was missing](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Controller/CRUDController.php#L1524) and now it return a Response, so [it doesn't execute some code](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Controller/CRUDController.php#L340).

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (franmomu) [reply src/Controller/CRUDController.php]: Technically, it should say `AdminInterface::getShow`, but since everyone extends from `AbstractAdmin`, it refers to the `AbstractAdmin::configureShowFields` where you define your fields. The same happen to [createAction](https://github.com/sonata-project/SonataAdminBundle/blob/master/src/Controller/CRUDController.php#L538) or editAction.

Comment (phansys) [reply src/Controller/CRUDController.php]: `AbstractAdmin::configureShowFields()` (and the others) are not abstract, so they aren't forcing to override their definition :thinking:

Comment (franmomu) [reply src/Controller/CRUDController.php]: Yes, you're right, they are not forced to override it, only when they want to access to the action. In the docs [we are telling people to extend from `AbstractAdmin`](https://sonata-project.org/bundles/admin/3-x/doc/getting_started/creating_an_admin.html#step-1-create-an-admin-class) and in this case, for the `show` action, [we are telling them to override `configureShowFields`](https://sonata-project.org/bundles/admin/3-x/doc/reference/action_show.html#the-show-action), so I think the developer would know what `configureShowFields` method means, maybe I should change it to something like:
`No fields defined, Did you forget to override the configureShowFields  in your Admin class?`.

File: src/Controller/CRUDController.php [hunk 5]
```diff
@@ -881,29 +827,10 @@ public function exportAction(Request $request)
...
-                $format
-            );
-            $exporter = $this->get('sonata.admin.exporter');
-        } else {
-            $adminExporter = $this->get('sonata.admin.admin_exporter');
-            $allowedExportFormats = $adminExporter->getAvailableFormats($this->admin);
-            $filename = $adminExporter->getExportFilename($this->admin, $format);
-            $exporter = $this->get('sonata.exporter.exporter');
-        }
+        $adminExporter = $this->get('sonata.admin.admin_exporter');
```
Comment (core23) [inline src/Controller/CRUDController.php]: Not part of this PR, but we should move this to the constructor or the method parameters

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (phansys) [reply src/Controller/CRUDController.php]: It's fine for me in the context of this PR, since IMO, this situation deserves a more exhaustive analysis.

File: tests/Controller/CRUDControllerTest.php [hunk 6]
```diff
@@ -196,16 +196,9 @@ protected function setUp(): void
             ->method('render')
             ->will($templatingRenderReturnCallback);
 
-        // NEXT_MAJOR : require sonata/exporter ^1.7 and remove conditional
-        if (class_exists(Exporter::class)) {
-            $exporter = new Exporter([new JsonWriter('/tmp/sonataadmin/export.json')]);
-        } else {
-            $exporter = $this->createMock(SonataExporter::class);
+        $exporter = new Exporter([new JsonWriter('/tmp/sonataadmin/export.json')]);
```
Comment (phansys) [inline tests/Controller/CRUDControllerTest.php]:
```suggestion
        $exporter = new Exporter([new JsonWriter(sys_get_temp_dir().'/sonataadmin/export.json')]);
```

File: src/Controller/CRUDController.php [hunk 7]
```diff
@@ -1521,12 +1429,10 @@ private function setFormTheme(FormView $formView, array $theme = null): void
...
     }
 
-    private function handleXmlHttpRequestErrorResponse(Request $request, FormInterface $form): ?JsonResponse
+    private function handleXmlHttpRequestErrorResponse(Request $request, FormInterface $form): JsonResponse
     {
         if ('application/json' !== $request->headers->get('Accept')) {
-            @trigger_error('In next major version response will return 406 NOT ACCEPTABLE without `Accept: application/json`', E_USER_DEPRECATED);
-
-            return null;
+            return $this->renderJson([], Response::HTTP_NOT_ACCEPTABLE);
```
Comment (phansys) [inline src/Controller/CRUDController.php]: Maybe this change should be made in the stable branch first, but the `Accept` header could allow "application/json" and at the same time other types:
>application/json, application/xml;q=0.9, \*/\*;q=0.8

Even, it could include extra data, like API versioning:
>application/json;version=2.0

Comment (phansys) [reply src/Controller/CRUDController.php]: I've created #5962 to solve this.

**Comment by greg0ire** [comment]
Thanks @franmomu !
